### DISCUSSION TITLE
TODOTXT - a related concept? why not join forces?

### DISCUSSION THREAD

**Discussion by nnako** [GothenburgBitFactory/taskwarrior#2662] [General]
Hi,

about 2015 I started developing my own task management application. Unfortunately, at that time, I did somehow overlook the TaskWarrior concept. So, I built my stuff around the TODOTXT concept which (as well) uses plaintext-based file representations of todo elements, flexible attributes, special attributes, project names, task priorities (A..Z), ...

Can anybody tell me, why there must be these two "similar" concepts TASKWARRIOR and TODOTXT? For both, there seems to be a huge user community, loads of applications / tools for all kinds of user interfacing and synchronization. And in both concepts, the `collaboration` feature has not successfully been dealt with, yet.

Now, from a TaskWarrior perspective, some questions:

- **What are the general benefits of TaskWarrior over the TODOTXT ecosystem?**
  of course, there are features like `urgency calculation`, `builtin, flexible reports`,
  `until dates`, `filters`, ... which provide a "better" usability on the task level. But
  besides features?
  
- **What are the key technical differences between these two concepts?**
  As I have heard, TaskWarrior is implemented in C++ and TODOTXT was originally
  written in BASH. Both with their tools implemented in various languages. But would
  there be things which couldn't be realized with one of the two?
  
- **Would it be possible (at least theoretically) to join these two projects?**
  Join forces in the way that each developer does not have to decide for which
  project to invest his time? As both operate on a per-person-level (currently
  at least) and use plain-text data storage (with a different syntax / format, though),
  even a simple file format converter would be suitable enough to combine the
  concepts, though of course, the features would be only applicable respective to
  the chosen concept. But with tasks interchangible.
  
- **What are the obstacles to joining the projects?**
  As I don't think to have been the first or only one thinking about TODOTXT here,
  maybe there are good reasons which I can not see, yet? Or maybe the join has
  already taken place and I didn't see it?

Hope for an interesting discussion, here.

Thanks.

**Comment by smemsh** [comment]
well, the projects are quite different, and have different histories, codebases, data formats, etc.  There are many different task management tools.  Probably hundreds if you include web-based software.  Good to have choices.  I'm sure there's import/export tools from/to taskwarrior/todotxt.  Here's one I found: https://github.com/juzim/taskwarrior2todo.txt

In theory, one could write a taskwarrior hook to insert/modify/delete tasks in both places.   Perhaps todo.txt has hooks also, you could use the hooks on both sides, if you really wanted to use two tools on the same data.  But you could also just choose one tool or the other based on which you liked better... I don't know of any particular reason to join these tools.

**Comment by nnako** [comment]
Hi @smemsh . Thanks for your reply.

In a way, you are right: most task management systems are different concerning development histories, codebases, internal data formats, data storage formats, data privacy, implementation language, UI concept, ... although on a higher level, they seem to very similar as they all have one goal: to be able to manage tasks ;-) . Sometimes, I have get impression, *every* junior developer creates his own task management application ... and this is too bad, as it leads to "wasted" time and effort.

From all the community-developed task management applications I saw so far, TaskWarrior and TODOTXT seem to be having a big user community and a similar usage concept:

- using local (private) text files as their main data storage (with different formats, though)
- having a command line interface as their main UI for task management
- further tools access these text files and extend the core functionality

but, yes, there are plenty of differences under the hood.

The reason for me starting the discussion was to get closer to a personal decision about possibly changing sides and investing some of my time in contributing to the TaskWarrior ecosystem. As I have already developed my own multi-user task distribution / management / notification application based on an extended TODOTXT file format, the usage of NextCloud for synchronization and simple email for notification, I am looking for a suitable way to possibly integrate it into an existing project rather than creating and maintaining my own task management island and having to add features like urgency management myself.

Operating both systems at the same time by keeping their text files in sync does not seem to be an option to me as I would have to take care of the textual representation of all the features within both systems in order not to break any functionality of either system.

Does this make sense?

**Reply by sl4mmy** [reply]
It sounds like you are simply asking whether or not anyone has already built any integration tools for taskwarrior and TODOTXT, is that correct?

In your original message when you talked about "joining the projects" I assumed you meant merging the two tools & their respective communities into one single project.  From your follow-up message in response to @smemsh it sounds like that's not what you're asking at all, and you would even be satisfied with just two data migration scripts: one to import TODOTXT data into taskwarrior, and one to import taskwarrior data into TODOTXT.

**Reply by nnako** [reply]
Hi @sl4mmy , thanks for your interpretation ;-)

> It sounds like you are simply asking whether or not anyone
> has already built any integration tools for taskwarrior and
> TODOTXT, is that correct?

I wanted to express my wish about proceeding towards merging the two (in my view) most popular open source, plain-text-file-based and command line operated solutions into one application. I see that developers invest time and effort into each of these projects. If it was possible to merge them, features could be implemented and become available for bothe user communities. Not just integrating one into the other but merging the projects. Might be that I am the only one thinking this would be a good thing to do. This is why I asked wheter or not there would be major obstacles in proceeding into that direction.

There are projects out there which have split their user community. With good reasons to do so, of course. As I see it, splitting up was not rooted in technological but rather in strategical differences:

- Owncloud -> Nextcloud
- FreeMind -> Freeplane
- OpenOffice -> LibreOFfice
- ...

So, if there were no fundamental non-tech reasons, why not provide the user with *one* application that has all the features from both projects built in? For me personally, I want to contribute to *one* project. If a merge of projects would be pursued, I would be happy to contribute to the merge. If both projects stay separate, I would have to decide where to contribute. That was the background of my question.

> ... From your follow-up message in response to @smemsh it
> sounds like that's not what you're asking at all, and you would
> even be satisfied with just two data migration scripts: one to
> import TODOTXT data into taskwarrior, and one to import
> taskwarrior data into TODOTXT.

Sorry for the confusion. Being able to integrate one text file into the other system could be seen as a first step towards a merge,of the projects, yes. At least to start making available the intersection features of both, TaskWarrior and TODOTXT to either user communities. But e.g. in order to use the urgency calculation feature, the user's tasks must be in TaskWarrior text format and the evaluation must take place with the TaskWarrior application. As long as there is no extension to TODOTXT which performs comparable calculations and visualizations directly on it's native format.

**Reply by tbabej** [reply]
> So, if there were no fundamental non-tech reasons, why not provide the user with one application that has all the features from both projects built in? For me personally, I want to contribute to one project. If a merge of projects would be pursued, I would be happy to contribute to the merge. If both projects stay separate, I would have to decide where to contribute. That was the background of my question.

Hi @nnako , perhaps I can provide some clarity on this part :slightly_smiling_face: In general, I am all for joining forces in the space - there is always a shortage of contributors, and the list of improvements is never decreasing. However, in this particular case, I'd say the short answer is that you'll have to make a choice as to where you'd like to contribute, as the merger of the projects at this point is relatively unlikely, given they both existed in isolation for more than decade and a half. I'd also point out, although it's not a great metric, that just going by the commit activity, todo.txt is currently less actively developed (~100 commits in the last decade for [the main client](https://github.com/todotxt/todo.txt-cli/graphs/contributors), vs. ~7000 for taskwarrior), so it's unclear if there are a lot of forces to join in the first place :thinking:

**Reply by nnako** [reply]
Hi @tbabej , and thanks for your reply.

I see, `todo.txt` does not seem to be so active anymore, especially, when looking at only the CLI application itself. There have been some efforts in the past, though, which definitely resulted in various applications for user interfacing (simple, curses-based, Android-based, ...). But, yes, seeing the corresponding [contribution page](https://github.com/GothenburgBitFactory/taskwarrior/graphs/contributors) for `TaskWarrior`, I am encouraged to becoming active. Anyway...

I think, I might try to get acquainted to the `TaskWarrior` ecosystem and discover its features and development structure. Hopefully, I will be succesful in this so that contribution will succeed.

Thanks.
